File: sdks/python/src/bounded_authority_verifier/facts.py
"""Value-bearing, redacted, non-authorizing facts (protocol-v1.md § Public verification contract,
REQ1-VERIFY-facts-redacted, REQ1-VERIFY-facts-not-credentials).

No generic encoder, no ``asdict``, no authorization/decision method (AGENTS rule 1). Grant/Envelope/
Export carry ``authorization="not_evaluated"``; Chain/Anchor/Transition carry
``trust="not_evaluated"``.

These are frozen dataclasses — value objects. They expose their fields for inspection but carry no
behavior that converts them into a credential or a decision.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class GrantFacts:
    """GrantFacts (protocol-v1.md:322-324): version, issuer, grant_id, issuer-key fingerprint (raw 32),
    holder thumbprint (raw 32), matched audience, grant times, authorization=not_evaluated."""

    version: int
    issuer: str
    grant_id: str
    issuer_key_fingerprint: bytes
    holder_thumbprint: bytes
    matched_audience: str
    issued_at: int
    not_before: int
    expires_at: int
    authorization: str = "not_evaluated"


@dataclass(frozen=True)
class EnvelopeFacts:
    """EnvelopeFacts (protocol-v1.md:342-348): GrantFacts + proof_id, invocation_id, operation, uri,
    grant hash (ath raw 32), request hash (ba_req raw 32), proof issued_at."""

    version: int
    issuer: str
    grant_id: str
    issuer_key_fingerprint: bytes
    holder_thumbprint: bytes
    matched_audience: str
    grant_issued_at: int
    grant_not_before: int
    grant_expires_at: int
    proof_id: str
    invocation_id: str
    operation: str
    uri: str
    grant_hash: bytes
    request_hash: bytes
    proof_issued_at: int
    authorization: str = "not_evaluated"


@dataclass(frozen=True)
class ChainFacts:
    """ChainFacts (ADR 0004:84-87; REQ1-CHAIN-facts-shape): value-bearing, trust=not_evaluated.

    Field set mirrors the Elixir reference ChainFacts (chain_facts.ex) exactly.
    """

    version: int
    chain_id: str
    first_sequence: int
    last_sequence: int
    row_count: int
    previous_hash: bytes
    last_hash: bytes
    verification: str = "boundary_consistent"
    trust: str = "not_evaluated"


@dataclass(frozen=True)
class AnchorFacts:
    """AnchorFacts (protocol-v1.md § Historical anchor verify): trust=not_evaluated.

    Field set mirrors the Elixir reference AnchorFacts (anchor_facts.ex) exactly. The reference
    carries NO key_id (only the key_fingerprint).
    """

    version: int
    anchor_id: str
    anchored_at: int
    chain_id: str
    sequence: int
    chain_hash: bytes
    key_fingerprint: bytes
    verification: str = "signature_and_window"
    trust: str = "not_evaluated"


@dataclass(frozen=True)
class KeyTransitionFacts:
    """KeyTransitionFacts (ADR 0004:44-55): trust=not_evaluated.

    Field set mirrors the Elixir reference KeyTransitionFacts (key_transition_facts.ex) exactly. The
    reference carries fingerprints only (current_/next_key_fingerprint), no key ids.
    """

    version: int
    transition_id: str
    chain_id: str
    effective_at: int
    current_key_fingerprint: bytes
    next_key_fingerprint: bytes
    verification: str = "authenticated_transition"
    trust: str = "not_evaluated"


@dataclass(frozen=True)
class AnchoredExportFacts:
    """AnchoredExportFacts (protocol-v1.md:437-440): the ONLY facts type with an authorization field.

    Field set mirrors the Elixir reference AnchoredExportFacts (anchored_export_facts.ex) exactly.
    """

    version: int
    chain_id: str
    first_sequence: int
    last_sequence: int
    row_count: int
    previous_hash: bytes
    last_hash: bytes
    digest: bytes
    start_anchor_id: str
    start_anchored_at: int
    start_key_fingerprint: bytes
    end_anchor_id: str
    end_anchored_at: int
    end_key_fingerprint: bytes
    transition_count: int
    object_version: str
    verification: str = "anchored_export"
    trust: str = "not_evaluated"
    authorization: str = "not_evaluated"


@dataclass(frozen=True)
class GrantDecoded:
    """Decode surface — structural grant facts (verification=not_evaluated)."""

    key_id: str
    issuer: str
    grant_id: str
    audiences: tuple[str, ...]
    issued_at: int
    not_before: int
    expires_at: int
    holder_thumbprint: bytes
    verification: str = "not_evaluated"


@dataclass(frozen=True)
class ProofDecoded:
    """Decode surface — structural proof facts (verification=not_evaluated)."""

    proof_id: str
    holder_thumbprint: bytes
    verification: str = "not_evaluated"


@dataclass(frozen=True)
class KeyLocator:
    """Locator surface — header-only key id (trust=not_evaluated)."""

    key_id: str
    trust: str = "not_evaluated"

File: sdks/go/facts.go
package verifier

// Verified facts (docs/protocol-v1.md § Public verification contract). Facts
// are value-bearing and redacted: they contain only their exact documented
// fields, no raw credentials, signatures, JWK containers, nonces, request
// argument values, or selector values. They carry no JSON, string, or generic
// encoder and are not credentials (REQ1-VERIFY-facts-redacted,
// REQ1-VERIFY-facts-not-credentials). Go cannot seal %+v formatting the way
// the Elixir seals Inspect; the redaction contract is the closed field set
// itself.

// Authorization is the fixed non-authorizing marker on GrantFacts,
// EnvelopeFacts, and AnchoredExportFacts. Only NotEvaluated exists.
type Authorization int

const (
	AuthorizationNotEvaluated Authorization = iota
)

// Trust is the fixed not-a-trust-decision marker on the diagnostic chain,
// anchor, transition, and export facts. Only NotEvaluated exists.
type Trust int

const (
	TrustNotEvaluated Trust = iota
)

// GrantFacts is the standalone raw-grant verification result.
type GrantFacts struct {
	Version              int
	Issuer               string
	GrantID              string
	IssuerKeyFingerprint [32]byte
	HolderThumbprint     [32]byte
	MatchedAudience      string
	IssuedAt             int64
	NotBefore            int64
	ExpiresAt            int64
	Authorization        Authorization
}

// EnvelopeFacts is the combined grant+proof verification result.
type EnvelopeFacts struct {
	Version              int
	Issuer               string
	GrantID              string
	IssuerKeyFingerprint [32]byte
	HolderThumbprint     [32]byte
	MatchedAudience      string
	IssuedAt             int64
	NotBefore            int64
	ExpiresAt            int64
	ProofID              string
	InvocationID         string
	Operation            string
	TargetURI            string
	GrantHash            [32]byte
	RequestHash          [32]byte
	ProofIssuedAt        int64
	Authorization        Authorization
}

// ChainFacts is the consumption-chain range verification result: bounded
// identifiers, times, counts, hashes, and performed-verification labels only
// (REQ1-CHAIN-facts-shape).
type ChainFacts struct {
	ChainID       string
	FirstSequence int64
	LastSequence  int64
	RowCount      int64
	LastHash      [32]byte
	Checks        []string
	Trust         Trust
}

// AnchorFacts is the historical boundary-anchor verification result.
type AnchorFacts struct {
	AnchorID   string
	ChainID    string
	KeyID      string
	Sequence   int64
	AnchoredAt int64
	Trust      Trust
}

// KeyTransitionFacts is the historical key-transition verification result.
type KeyTransitionFacts struct {
	TransitionID    string
	ChainID         string
	EffectiveAt     int64
	FromFingerprint [32]byte
	ToFingerprint   [32]byte
	ToKeyID         string
	Trust           Trust
}

// AnchoredExportFacts is the complete archive verification result. It states
// the performed cryptographic checks and never carries rows, commitments,
// compacts, archive bytes, or the object-store version.
type AnchoredExportFacts struct {
	ChainID         string
	FirstSequence   int64
	LastSequence    int64
	RowCount        int64
	TransitionCount int
	ChunkCount      int
	ByteCount       int64
	Digest          [32]byte
	Checks          []string
	Trust           Trust
	Authorization   Authorization
}

// KeyLocator is the untrusted-key-locator result: a closed kid hint with
// trust not evaluated (REQ1-LOCATOR-not-authority).
type KeyLocator struct {
	KeyID string
	Trust Trust
}
